zerver/actions/scheduled_messages.py

```diff
@@ -145,33 +143,111 @@ def notify_update_scheduled_message(
-        scheduled_message_object.set_topic_name(topic_name=topic_name)
-        rendering_result = render_markdown(
-            send_request.message, send_request.message.content, send_request.realm
-        )
-        scheduled_message_object.content = send_request.message.content
-        scheduled_message_object.rendered_content = rendering_result.rendered_content
-        scheduled_message_object.sending_client = send_request.message.sending_client
-        scheduled_message_object.stream = send_request.stream
-        assert send_request.deliver_at is not None
-        scheduled_message_object.scheduled_timestamp = send_request.deliver_at
+        # If the server failed to send the scheduled message, a new scheduled
+        # delivery timestamp (`deliver_at`) is required.
+        if scheduled_message_object.failed and deliver_at is None:
+            raise JsonableError(_("Scheduled delivery time must be in the future."))
 
-        scheduled_message_object.has_attachment = check_attachment_reference_change(
-            scheduled_message_object, rendering_result
+        # Get existing scheduled message's recipient IDs and recipient_type_name.
+        existing_recipient, existing_recipient_type_name = get_recipient_ids(
+            scheduled_message_object.recipient, sender.id
         )
 
+        # If any recipient information or message content has been updated,
+        # we check the message again.
+        if recipient_type_name is not None or message_to is not None or message_content is not None:
+            # Update message type if changed.
+            if recipient_type_name is not None:
+                updated_recipient_type_name = recipient_type_name
+            else:
+                updated_recipient_type_name = existing_recipient_type_name
+
+            # Update message recipient if changed.
+            if message_to is not None:
+                if updated_recipient_type_name == "stream":
+                    updated_recipient = extract_stream_id(message_to)
+                else:
+                    updated_recipient = extract_direct_message_recipient_ids(message_to)
+            else:
+                updated_recipient = existing_recipient
+
+            # Update topic name if changed.
+            if topic_name is not None:
+                updated_topic = topic_name
+            else:
+                # This will be ignored in Addressee.legacy_build if type
+                # is being changed from stream to direct.
+                updated_topic = scheduled_message_object.topic_name()
+
+            # Update message content if changed.
+            if message_content is not None:
+                updated_content = message_content
+            else:
+                updated_content = scheduled_message_object.content
+
+            # Check message again.
+            addressee = Addressee.legacy_build(
+                sender, updated_recipient_type_name, updated_recipient, updated_topic
+            )
```

Comment:
I suspect, but am not certain, that this could be extracted into a function that can be shared by the `create_scheduled_message_backend` code that returns an object "Here's the requested content, and the requested recipient type and recipient". I'm not certain that would be cleaner, but it might be.